x/stakeibc/keeper/callbacks.go

```diff
@@ -61,123 +60,127 @@ func (c Callbacks) RegisterCallbacks() icqtypes.QueryCallbacks {
 // WithdrawalBalanceCallback is a callback handler for WithdrawalBalance queries.
 func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icqtypes.Query) error {
 	// NOTE(TEST-112) for now, to get proofs in your ICQs, you need to query the entire store on the host zone! e.g. "store/bank/key"
-	k.Logger(ctx).Error(fmt.Sprintf("WithdrawalBalanceCallback: %v", query))
+	k.Logger(ctx).Info(fmt.Sprintf("WithdrawalBalanceCallback executing, QueryId: %vs, Host: %s, QueryType: %s, Height: %d, Connection: %s",
+		query.Id, query.ChainId, query.QueryType, query.Height, query.ConnectionId))
 
-	zone, found := k.GetHostZone(ctx, query.GetChainId())
+	hostZone, found := k.GetHostZone(ctx, query.GetChainId())
 	if !found {
-		return fmt.Errorf("no registered zone for chain id: %s", query.GetChainId())
+		errMsg := fmt.Sprintf("no registered zone for queried chain ID (%s)", query.GetChainId())
+		k.Logger(ctx).Error(errMsg)
+		return sdkerrors.Wrapf(types.ErrHostZoneNotFound, errMsg)
 	}
-	balancesStore := query.Request[1:]
-	accAddr, err := banktypes.AddressFromBalancesStore(balancesStore)
+
+	// Query request is a byte array of the form:
+	// [ {balancesPrefix} {address} {denom} ]
+	// {balancePrefix} is only a single byte - and it must be removed before calling AddressFromBalancesStore
+	balancesStoreKey := query.Request[1:]
+	queriedAddress, err := banktypes.AddressFromBalancesStore(balancesStoreKey)
 	if err != nil {
-		return err
+		errMsg := fmt.Sprintf("unable to derive queried address from request byte array")
+		k.Logger(ctx).Error(errMsg)
+		return sdkerrors.Wrapf(err, errMsg)
 	}
 
 	//TODO(TEST-112) revisit this code, it's not vetted
-	coin := sdk.Coin{}
-	err = k.cdc.Unmarshal(args, &coin)
+	// Unmarshal the CB args into a coin type
+	withdrawalBalanceCoin := sdk.Coin{}
+	err = k.cdc.Unmarshal(args, &withdrawalBalanceCoin)
 	if err != nil {
-		k.Logger(ctx).Error(fmt.Sprintf("unable to unmarshal balance info for zone: %s, err: %s", zone.ChainId, err.Error()))
-		return err
+		errMsg := fmt.Sprintf("unable to unmarshal balance in callback args for zone: %s, err: %s", hostZone.ChainId, err.Error())
+		k.Logger(ctx).Error(errMsg)
+		return sdkerrors.Wrapf(types.ErrMarshalFailure, errMsg)
 	}
 
-	if coin.IsNil() {
-		denom := ""
-
-		for i := 0; i < len(query.Request)-len(accAddr); i++ {
-			if bytes.Equal(query.Request[i:i+len(accAddr)], accAddr) {
-				denom = string(query.Request[i+len(accAddr):])
-				break
-			}
-
-		}
-		// if balance is nil, the response sent back is nil, so we don't receive the denom. Override that now.
-		coin = sdk.NewCoin(denom, sdk.ZeroInt())
-	}
+	// SOMEONE DOUBLE CHECK ME ON THIS
+	// It looks unmarshalling a nil coin amount converts it to zero, which would mean this removed branch should be impossible
```

Comment:
I think it wouldn't matter though right? Cause if the amount is 0, we exit immediately after this

Reply:
I don't think it's possible for the coin itself to be null. That would imply the byte array we pass into this callback is nil, which I don't think is possible (or if it is, then we should just add a check upstream - maybe even in validate basic for submit_query).

I think it's likely that if it can't find a balance, it just returns sdk.Coin{}. Which will then have a 0 amount after unmarshalling and then we'll exit.

Reply:
Ah, didn't realize it cannot be nil, looks like the line 88 error catching prevents that. In that case I'm all signed off on this @sampocs! 🚢